drivers: add driver for FXOS8700 3-axis accelerometer/magnetometer

Hyungsin opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

This PR provides a driver for the FXOS8700 sensor, which reads 3-axis acceleration and magnetic field.
Driver implementation style is similar to HDC1000 driver.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

gebart commented:
Nice! This sensor can be found on the frdm-kw41z board.

gebart commented on drivers/fxos8700/include/fxos8700_params.h:
```diff
@@ -0,0 +1,75 @@
+ */
+static const saul_reg_info_t fxos8700_saul_info[] =
+{
+    HDC1000_SAUL_INFO
```
Copy pasta

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,203 @@
+        DEBUG("[fxos8700] Could not read WHOAMI (%d)\n", rv);
+        return FXOS8700_NOBUS;
+    }
+    dev->whoami = config;
```
Is there any use case for saving this value? It seems like it is never used by the implementation.

gebart commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,293 @@
+} fxos8700_params_t;
+
+/**
+  * @brief   Device descriptor for a AT30TSE75x device
```
Copy pasta, and weird indentation

gebart commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,293 @@
+} fxos8700_t;
+
+/**
+  * @brief   Individual hybrid measurement
```
Weird indent

gebart commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,293 @@
+        uint8_t  ydr:1;       /*!< bit:      0  Start Conversion Event In          */
+        uint8_t  xdr:1;       /*!< bit:      1  Synchronization Event In           */
+    } bit;                    /*!< Structure used for bit  access                  */
+    uint8_t reg;              /*!< Type      used for register access              */
```
Here and everywhere else: use `/**` (and of course `/**<`) for doxygen comments in RIOT.

gebart commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,293 @@
+        uint8_t  xdr:1;       /*!< bit:      1  Synchronization Event In           */
+    } bit;                    /*!< Structure used for bit  access                  */
+    uint8_t reg;              /*!< Type      used for register access              */
+} FXOS8700_STATUS_Type;
```
This type, and the other bit fields below, are not actually used by the implementation.
Additionally, it is not required by a compiler to pack the bit fields in the struct, so this bit representation will not be guaranteed to match the hardware register bits. It is more robust to use bit mask macros or other bit operations in the implementation. The compiler should be smart enough to optimize the accesses in a good way when one of the operands is a numeric literal or static constant.

gebart commented:
The cache does not handle multiple instances.
I understand that your projects use the caching to reduce the power consumption while reading multiple successive reads, but I wonder if we should generalize the cache into its own module and let it wrap any sensor driver, with configurable cache time to live. Most sensors have the same kind of read function API where you pass a driver instance as the first argument and an output buffer as the next argument with error codes as the return value, that would be easily wrapped by a generic cache implementation.

gebart commented:
Otherwise we will likely end up with various cache implementations in each sensor driver, with plenty of opportunity for mistakes and difficulties testing them in a reliable way.

gebart commented:
I think it is a good verification to read the whoami register during init, to check that the device is responding and that the connected device is actually a FXOS8700 sensor by comparing the actual value against the expected value. Just that it does not need to be saved in the context struct.

Hyungsin commented:
@gebart. Thank you for the quick review! I resolved you comments.

WHOAMI comparison is still here but its value is not saved now.

Regarding the cache, I don't know how many sensors in RIOT are relevant (one sensor should provide multiple types of information). If this cache functionality is useful for many other sensors, we can think of generalizing it. But I think this should be another PR.

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+        int32_t acc_raw_z = (int32_t) ((data[4]<<6) | (data[5]>>2));
+        acc->x = (int16_t)(acc_raw_x*244/1000);
+        acc->y = (int16_t)(acc_raw_y*244/1000);
+        acc->z = (int16_t)(acc_raw_z*244/1000);
```
Code style: add spaces around operators

gebart replied:
Also, for clarity it may be better to have parentheses around `(raw * 244) / 1000`, though it changes nothing in the generated code since the associativity is left to right for the multiplication and division operators.

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+    if (mag) {
+        mag->x = (int16_t) ((data[6] <<8) | data[7]);
+        mag->y = (int16_t) ((data[8] <<8) | data[9]);
+        mag->z = (int16_t) ((data[10]<<8) | data[11]);
```
Spaces around bit shift operators

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+
+int fxos8700_read_cached(const fxos8700_t *dev, fxos8700_measurement_t* acc, fxos8700_measurement_t* mag)
+{
+    uint32_t now = xtimer_now_usec();
```
Optional: It may be less computationally expensive to use xtimer_now and convert the renew interval to ticks instead, since you can precompute that us to tick conversion at compilation time rather than do a conversion from ticks to us in every read. See this as an optional comment, and maybe just remember for future drivers, I assume this function will be called somewhat infrequently since the renew interval is set to 1 second.

gebart commented:
Are you using this sensor externally or integrated on any dev board? Do you have any updated board configurations you want to add here?

Hyungsin commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+    i2c_acquire(dev->p.i2c);
+    if(i2c_read_regs(dev->p.i2c, dev->p.addr, reg, (char*) data, len) <= 0) {
+        DEBUG("[fxos8700] Can't read register 0x%x\n", reg);
+        i2c_release(dev->p.i2c);
```
I followed the style of the current HDC1000 implementation. Is this style a strict requirement?

aabadie replied:
No, that was more a suggestion for future maintainance, technically it doesn't change anything.

Hyungsin replied:
I see. Thanks!

Hyungsin commented:
@gebart @aabadie 
Thank you for the reviews and I resolved comments again. 👍 

@gebart, I'm using FXOS8700 on the hamilton board, which I plan to include in RIOT.
Basically, all sensor/clock/radio drivers I have modified/added so far are related to the hamilton.

aabadie commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+#define MAG_ONLY_MODE           (0x01)
+#define HYBRID_MODE             (0x03)
+
+static fxos8700_measurement_t acc_cached, mag_cached;
```
Instead of those static variable; why no having them as attributes of `fxos8700_t` ?
Because otherwise this will be a problem with multiple instances of this device as already reported by @gebart [here](https://github.com/RIOT-OS/RIOT/pull/8978#issuecomment-382612342).That could be a best temporary solution until we have a module for managing cache.
Maybe saul could do it ?

Hyungsin replied:
I agree and updated this PR reflecting this comment. Thanks!

aabadie commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,277 @@
+ * @brief   Device descriptor for a FXOS8700 device
+ */
+typedef struct {
+    fxos8700_measurement_t acc_cached;
```
The attributes of this struct are missing documentation, same with the fxos8700_measurement_t struct

aabadie commented:
Could you also add a test application ?

Hyungsin commented:
@aabadie, sure. I added an application. Thanks!

Hyungsin commented:
Now all checks have passed!

aabadie commented on tests/driver_fxos8700/Makefile:
```diff
@@ -0,0 +1,7 @@
+include ../Makefile.tests_common
+
+BOARD_BLACKLIST := msb-430 msb-430h telosb wsn430-v1_3b wsn430-v1_4 z1
```
is this needed ?

aabadie commented on tests/driver_fxos8700/README.md:
```diff
@@ -0,0 +1,12 @@
+If you want to use this test application with different parameters, you can
+simply override the default FXOS8700_PARAMS.
+
+After initialization, the sensor reads the pulse counter values every second and
```
copy-paste (remaining occurrence of pulse counter)

aabadie commented on tests/driver_fxos8700/main.c:
```diff
@@ -0,0 +1,56 @@
+        fxos8700_measurement_t acc, mag;
+        if (fxos8700_read(&dev, &acc, &mag) != FXOS8700_OK) {
+            printf("Read failure\n");
+        } else {
```
the else should on next line

aabadie commented on tests/driver_fxos8700/main.c:
```diff
@@ -0,0 +1,56 @@
+{
+    fxos8700_t dev;
+
+    printf("FXOS8700 driver test application\n");
```
maybe use puts instead of printf

aabadie commented on tests/driver_fxos8700/main.c:
```diff
@@ -0,0 +1,56 @@
+
+    /* Initialization */
+    if (fxos8700_init(&dev, &fxos8700_params[0])) {
+        printf("[Failed]");
```
'puts' here and in other places below

aabadie commented on tests/driver_fxos8700/main.c:
```diff
@@ -0,0 +1,56 @@
+            printf("acceleration: (%d, %d, %d)\n", acc.x, acc.y, acc.z);
+            printf("magnetic field: (%d, %d, %d)\n", mag.x, mag.y, mag.z);
+        }
+        xtimer_usleep(SLEEP);
```
since you are waiting 1s, you can use xtimer_sleep with 1 as parameter. Then you can also drop the define SLEEP above

aabadie commented on drivers/include/fxos8700.h:
```diff
@@ -0,0 +1,277 @@
+#endif
+
+
+/**
```
Maybe put those defines in a separate internal header file (see other drivers)

aabadie commented on tests/driver_fxos8700/main.c:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2017 UC Berkeley
```
2018

gebart commented:
I think something is wrong with the scale configuration. This is the SAUL output I see when I have the board completely still on my desk:

```
2018-04-20 17:23:36,600 - INFO # Reading from #14 (fxos8700|SENSE_ACCEL)
2018-04-20 17:23:36,602 - INFO # Data:	[0]   3742 mg
2018-04-20 17:23:36,603 - INFO # 	[1]   3957 mg
2018-04-20 17:23:36,605 - INFO # 	[2]    962 mg
2018-04-20 17:23:36,605 - INFO # 
2018-04-20 17:23:36,607 - INFO # Reading from #15 (fxos8700|SENSE_MAG)
2018-04-20 17:23:36,609 - INFO # Data:	[0]    763 mGs
2018-04-20 17:23:36,611 - INFO # 	[1]    271 mGs
2018-04-20 17:23:36,612 - INFO # 	[2]    683 mGs
2018-04-20 17:23:36,613 - INFO # 
```

The length (2-norm) of the acceleration vector should be close to 1000 mg, but we have some 5500 mg

gebart commented:
(or my sensor is broken)

Hyungsin commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,204 @@
+
+static int fxos8700_write_regs(const fxos8700_t* dev, uint8_t reg, uint8_t* data, size_t len)
+{
+    i2c_acquire(dev->p.i2c);
```
i2c_release is added now :)

Hyungsin commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,204 @@
+    /* Configure the ODR to maximum (400Hz in hybrid mode) */
+    config = 0x00;
+    if (fxos8700_write_regs(dev, FXOS8700_REG_CTRL_REG1, &config, 1) != FXOS8700_OK) {
+        return FXOS8700_BUSERR;
```
fxos8700_write_regs and fxos8700_read_regs have their own i2c_release.

Hyungsin commented:
@aabadie, @gebart, I addressed your comments.
and @gebart, let me test this with my device also and I will let you know. It worked before but there could be some bugs while rearranging codes.

Hyungsin commented:
@gebart, I think it works at least in my case (hamilton).
-------------------------------------------
Dev: fxos8700	Type: SENSE_ACCEL
Data:	[0] 15mg
        	[1] 1005mg
	        [2] 156mg
-------------------------------------------
Can you retry?

gebart commented:
I will check again tomorrow

gebart commented:
My frdm-kw41z is giving strange values, I will try again with a frdm-k22f board tomorrow

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    /* Read all data at once */
+    if (fxos8700_read_regs(dev, FXOS8700_REG_OUT_X_MSB, &data[0], 12)) {
```
change magic `12` into sizeof something or a macro definition.

gebart commented on drivers/fxos8700/fxos8700.c:
```diff
@@ -0,0 +1,205 @@
+    if (acc) {
+        int32_t acc_raw_x = (int32_t) ((data[0] << 6) | (data[1] >> 2));
+        int32_t acc_raw_y = (int32_t) ((data[2] << 6) | (data[3] >> 2));
+        int32_t acc_raw_z = (int32_t) ((data[4] << 6) | (data[5] >> 2));
```
the shift of 6 loses the sign bit

Hyungsin replied:
Very good point. Thanks!!

Hyungsin commented:
@gebart, I updated, but I think this driver does not lose any bit regardless of resolution mode since it provides "FXOS8700_USE_ACC_RAW_VALUES" mode.

gebart commented:
:+1:

gebart commented:
@Hyungsin in future PRs, please create fixup/squash commits for changes you add after the initial review to make the review process easier. You can amend typos or other minor corrections, but for anything a little bigger, such as when you added the different scaling factor configurations in this PR, use a separate commit. Then squash the commits when the reviewer requests it. It makes it easier for the reviewer to know what has changed between iterations.

Hyungsin commented:
@gebart, good point. I should keep this in mind!

aabadie commented:
All green here. Let's merge.
